### ISSUE TITLE
Paramspace does not respect '.' in column names

### ISSUE THREAD

**Issue by psteinb** [snakemake/snakemake#1673]
**Snakemake version**
```
$ python --version                                       
Python 3.9.9
$ snakemake --version
7.6.2
```

**Describe the bug**
`Paramspace` does not honor the fact, that column names may contain `.`. They are replaced by `_2` inside `Paramspace.instance_pattern` calls.

**Logs**
```bash
$ python -m pytest tests/test_smk_paramspace.py -s -k six
============================================================================ test session starts ============================================================================
platform linux -- Python 3.9.9, pytest-7.1.2, pluggy-1.0.0
rootdir: /home/steinb95/vouchers/hzb/hartmann_beamline_inversion/sbibm_with_surrogates
plugins: hydra-core-1.2.0
collected 9 items / 8 deselected / 1 selected                                                                                                                               

tests/test_smk_paramspace.py ['/tmp/first=1/second=A/_2=A/_3=4/four=AA/_5=a/run.yaml', '/tmp/first=2/second=B/_2=B/_3=5/four=BA/_5=a/run.yaml', '/tmp/first=3/second=C/_2=C/_3=6/four=CA/_5=a/run.yaml']
F

================================================================================= FAILURES ==================================================================================
_______________________________________________________________ test_paramspace_pattern_six_cols_with_period ________________________________________________________________

df =    first second third.a  third.c four five_or_six.2
0      1      A       A        4   AA             a
1      2      B       B        5   BA             a
2      3      C       C        6   CA             a

    def test_paramspace_pattern_six_cols_with_period(df):
    
        df = pd.DataFrame.from_dict(
            {
                "first": [1, 2, 3],
                "second": "A B C".split(),
                "third.a": "A B C".split(),
                "third.c": [4, 5, 6],
                "four": "AA BA CA".split(),
                "five_or_six.2": 3 * ["a"],
            }
        )
        spc = Paramspace(df, param_sep="=")
        obs = expand("/tmp/{replacethis}/run.yaml", replacethis=spc.instance_patterns)
        assert len(obs) > 0
        assert len(obs) == len(df)
        print(obs)
>       assert "third" in obs[0]
E       AssertionError: assert 'third' in '/tmp/first=1/second=A/_2=A/_3=4/four=AA/_5=a/run.yaml'

tests/test_smk_paramspace.py:206: AssertionError
```

**Minimal example**
```py
def test_paramspace_pattern_six_cols_with_period(df):

    df = pd.DataFrame.from_dict(
        {
            "first": [1, 2, 3],
            "second": "A B C".split(),
            "third.a": "A B C".split(),
            "third.c": [4, 5, 6],
            "four": "AA BA CA".split(),
            "five_or_six.2": 3 * ["a"],
        }
    )
    spc = Paramspace(df, param_sep="=")
    obs = expand("/tmp/{replacethis}/run.yaml", replacethis=spc.instance_patterns)
    assert len(obs) > 0
    assert len(obs) == len(df)
    print(obs)
    assert "third" in obs[0] #assert throws
```

**Comment by psteinb** [comment]
Just tested snakemake `7.7.0` with python `3.10`, `3.9` and `3.8` and this bug occurs in all 3 of them.

**Comment by psteinb** [comment]
I invested some time in this. I must admit that code like [this](https://github.com/snakemake/snakemake/blob/bbed20f7c1de220353715838716e0de80e694b8e/snakemake/utils.py#L687):

```python
    @property
    def instance_patterns(self):
        """Iterator over all instances of the parameter space (dataframe rows),
        formatted as file patterns of the form column1~{value1}/column2~{value2}/...
        or of the provided custom pattern.
        """
        return (
            self.pattern.format(
                *(
                    self.param_sep.join(("{}", "{}")).format(name, value)
                    for name, value in row._asdict().items()
                )
            )
            for row in self.dataframe.itertuples(index=False)
        )
```
is really hard to debug. Or I simply don't know how to use `pdb`. In any case, I found the core reason for this problem. Again, illustrated in a unit test:
```python
def test_pd_dataframe_interrows_vs_itertuples():

    mockdict = {
        "first": [1, 2, 3],
        "second": "A B C".split(),
        "third.a": "A B C".split(),
        "third.c": [4, 5, 6],
        "four": "AA BA CA".split(),
        "five_or_six.2": 3 * ["a"],
    }

    df = pd.DataFrame.from_dict(mockdict)
    exp = sorted([k for k in mockdict.keys()])

    allrows = list(df.iterrows())
    firstrow = allrows[0]
    obs1 = sorted(firstrow[1].index)

    assert exp == obs1  # works like a charm

    alltupl = list(df.itertuples(index=False))
    firsttupl = alltupl[0]
    obs2 = sorted(firsttupl._fields)

    assert exp == obs2  # boom!
```
**tl;dr** `pd.DataFrame.itertuples` doesn't like `.` in column names. `pd.DataFrame.iterrows` appears to work like a charm.

My environment:
```
$ python --version                               
Python 3.10.4
$ snakemake --version           
7.7.0
$ python -c 'import pandas as pd; print(pd.__version__)' 
1.4.2
```

The `pandas` [documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.itertuples.html?highlight=itertupl) hints to this issue:
> The column names will be renamed to positional names if they are invalid Python identifiers, repeated, or start with an underscore.

I hence suggest to not use `itertuples` here. I'll prepare a PR.
